mcp/src/main/java/io/modelcontextprotocol/client/transport/WebSocketClientTransport.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2024 - 2024 the original author or authors.
+ */
+
+package io.modelcontextprotocol.client.transport;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.WebSocket;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.modelcontextprotocol.spec.McpClientTransport;
+import io.modelcontextprotocol.spec.McpSchema;
+import io.modelcontextprotocol.util.Assert;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Sinks;
+import reactor.util.retry.Retry;
+
+/**
+ * The WebSocket (WS) implementation of the
+ * {@link io.modelcontextprotocol.spec.McpTransport} that follows the MCP HTTP with WS
+ * transport specification, using Java's HttpClient.
+ *
+ * @author Aliaksei Darafeyeu
+ */
+public class WebSocketClientTransport implements McpClientTransport {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketClientTransport.class);
+
+	private final HttpClient httpClient;
+
+	private final ObjectMapper objectMapper;
+
+	private final URI uri;
+
+	private final AtomicReference<WebSocket> webSocketRef = new AtomicReference<>();
+
+	private final AtomicReference<TransportState> state = new AtomicReference<>(TransportState.DISCONNECTED);
+
+	private final Sinks.Many<Throwable> errorSink = Sinks.many().multicast().onBackpressureBuffer();
+
+	/**
+	 * The constructor for the WebSocketClientTransport.
+	 * @param uri the URI to connect to
+	 * @param clientBuilder the HttpClient builder
+	 * @param objectMapper the ObjectMapper for JSON serialization/deserialization
+	 */
+	WebSocketClientTransport(final URI uri, final HttpClient.Builder clientBuilder, final ObjectMapper objectMapper) {
+		this.uri = uri;
+		this.httpClient = clientBuilder.build();
+		this.objectMapper = objectMapper;
+	}
+
+	/**
+	 * Creates a new WebSocketClientTransport instance with the specified URI.
+	 * @param uri the URI to connect to
+	 * @return a new Builder instance
+	 */
+	public static Builder builder(final URI uri) {
+		return new Builder().uri(uri);
+	}
+
+	/**
+	 * The state of the Transport connection.
+	 */
+	public enum TransportState {
+
+		DISCONNECTED, CONNECTING, CONNECTED, CLOSED
+
+	}
+
+	/**
+	 * A builder for creating instances of WebSocketClientTransport.
+	 */
+	public static class Builder {
+
+		private URI uri;
+
+		private final HttpClient.Builder clientBuilder = HttpClient.newBuilder()
+			.version(HttpClient.Version.HTTP_1_1)
+			.connectTimeout(Duration.ofSeconds(10));
+
+		private ObjectMapper objectMapper = new ObjectMapper();
```

Comment:
Lots of object mappers defined in lots of files. It would seem like it might be time to build a Jackson module and lock in on one that can be used across the code base.